pcdsdevices/interface.py

```diff
@@ -0,0 +1,893 @@
+"""
+Module for defining bell-and-whistles movement features
+"""
+import time
+import fcntl
+import logging
+import numbers
+import signal
+import re
+from contextlib import contextmanager
+from pathlib import Path
+from threading import Thread, Event
+from types import SimpleNamespace, MethodType
+from weakref import WeakSet
+
+import yaml
+from bluesky.utils import ProgressBar
+from ophyd.device import Kind
+from ophyd.status import wait as status_wait
+
+from . import utils as util
+
+logger = logging.getLogger(__name__)
+engineering_mode = True
+
+OphydObject_whitelist = ["name", "connected", "check_value", "log"]
+BlueskyInterface_whitelist = ["trigger", "read", "describe", "stage",
+                              "unstage"]
+Device_whitelist = ["read_attrs", "configuration_attrs", "summary",
+                    "wait_for_connection", "stop", "get", "configure"]
+Signal_whitelist = ["value", "put"]
+Positioner_whitelist = ["settle_time", "timeout", "egu", "limits", "move",
+                        "position", "moving"]
+
+
+class BaseInterface:
+    """
+    Interface layer to attach to any Device for SLAC features.
+
+    This class defines an API and some defaults for filtering tab-completion
+    results for new users to avoid confusion. The API involves setting the
+    tab_whitelist attribute on any subclass of BaseInterface. When in
+    non-engineering mode, only elements on the whitelists will be displayed to
+    the user.
+
+    Attributes
+    ----------
+    tab_whitelist: list
+        list of string regex to show in autocomplete for non-engineering mode
+    """
+    tab_whitelist = (OphydObject_whitelist + BlueskyInterface_whitelist +
+                     Device_whitelist + Signal_whitelist +
+                     Positioner_whitelist)
+    _filtered_dir_cache = None
+
+    def __init_subclass__(self):
+        string_whitelist = []
+        for cls in self.mro():
+            if hasattr(cls, "tab_whitelist"):
+                string_whitelist.extend(cls.tab_whitelist)
+            if getattr(cls, "tab_component_names", False):
+                for cpt_name in cls.component_names:
+                    if getattr(cls, cpt_name).kind != Kind.omitted:
+                        string_whitelist.append(cpt_name)
+        self._tab_regex = re.compile("|".join(string_whitelist))
+
+    def __dir__(self):
+        if get_engineering_mode():
+            return super().__dir__()
+        elif self._filtered_dir_cache is None:
+            self._init_filtered_dir_cache()
+        return self._filtered_dir_cache
+
+    def _init_filtered_dir_cache(self):
+        self._filtered_dir_cache = self._get_filtered_tab_dir()
+
+    def _get_filtered_tab_dir(self):
+        filtered = []
```

Comment:
Unnecessary to refactor, but a list comprehension here would be clean + compact:
```python
return [elem
        for elem in super().__dir__()
        if self._tab_regex.fullmatch(elem)]
```

Reply:
`Unnecessary to refactor, but here's the code for you already refactored`

Reply:
In general I shy away from multi-line list comprehensions because they can quickly get ugly but this one seems pretty clean